lib/src/test/java/io/ably/lib/test/realtime/ConnectionManagerTest.java

```diff
@@ -417,4 +422,182 @@ public void onConnectionStateChanged(ConnectionStateChange state) {
+		try {
+			ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+			final AblyRealtime ably = new AblyRealtime(opts);
+			final long newTtl = 1000L;
+			final long newIdleInterval = 1000L;
+			/* We want this greater than newTtl + newIdleInterval */
+			final long intervalBeforeReconnecting = 3000L;
+			final String channelName = "test-reattach-after-ttl";
+			ably.connection.on(ConnectionEvent.connected, new ConnectionStateListener() {
+				@Override
+				public void onConnectionStateChanged(ConnectionStateChange state) {
+					try {
+						Field connectionStateField = ably.connection.connectionManager.getClass().getDeclaredField("connectionStateTtl");
+						connectionStateField.setAccessible(true);
+						connectionStateField.setLong(ably.connection.connectionManager, newTtl);
+						Field maxIdleField = ably.connection.connectionManager.getClass().getDeclaredField("maxIdleInterval");
+						maxIdleField.setAccessible(true);
+						maxIdleField.setLong(ably.connection.connectionManager, newIdleInterval);
+					} catch (NoSuchFieldException|IllegalAccessException e) {
+						fail("Unexpected exception in checking connectionStateTtl");
+					}
+				}
+			});
+
+			ConnectionWaiter connectionWaiter = new ConnectionWaiter(ably.connection);
+			connectionWaiter.waitFor(ConnectionState.connected);
+			final String firstConnectionId = ably.connection.id;
+
+			/* Attach to channel */
+			final Channel channel = ably.channels.get(channelName);
+			channel.once(ChannelState.attached, new ChannelStateListener() {
+				@Override
+				public void onChannelStateChanged(ChannelStateChange stateChange) {
+					System.out.println("Channel attached for the first time");
+					assertEquals("Channel is attached", stateChange.current, ChannelState.attached);
+				}
+			});
+			channel.attach();
+
+            /* Before disconnecting wait that newTtl + newIdleInterval has passed */
+            try {
+                Thread.sleep(intervalBeforeReconnecting);
+            } catch(InterruptedException e) {}
+			ably.connection.connectionManager.requestState(ConnectionState.disconnected);
+
+            /* Since newTtl + newIdleInterval has passed we expect the connection to go into a suspended state */
+            connectionWaiter.waitFor(ConnectionState.suspended);
+
+			ably.connect();
+			ably.connection.once(ConnectionEvent.connected, new ConnectionStateListener() {
+				@Override
+				public void onConnectionStateChanged(ConnectionStateChange state) {
+					String secondConnectionId = ably.connection.id;
+					assertNotNull(secondConnectionId);
+					assertNotEquals("connection has a different id", firstConnectionId, secondConnectionId);
+					channel.once(ChannelState.attached, new ChannelStateListener() {
+						@Override
+						public void onChannelStateChanged(ChannelStateChange stateChange) {
+							System.out.println("Channel reattached after a new connection has been established");
+							ably.close();
```

Comment:
How we do know the test ends at this point?  Surely we need a clause here to confirm that this test has finished i.e. could this test not complete before this, or am I not reading this correctly?